src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,489 @@
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo
+// structure into a public key and its corresponding algorithm.
+// TODO: add the other signatures flavors, as defined in common.go
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	pubKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := pubKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf("unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes the credential struct of the Delegated Credential.
+func (cred *Credential) marshal() ([]byte, error) {
+	ser := make([]byte, 4)
+	binary.BigEndian.PutUint32(ser, uint32(cred.ValidTime/time.Second))
+
+	var serAlgo [2]byte
+	binary.BigEndian.PutUint16(serAlgo[:], uint16(cred.expCertVerfAlgo))
+	ser = append(ser, serAlgo[:]...)
+
+	// Encode the public key
+	serPub, err := cred.marshalPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	// Assert that the public key encoding is no longer than 2^24  bytes.
+	if len(serPub) > dcMaxPubLen {
+		return nil, errors.New("tls: public key is not valid")
+	}
+
+	var serLen [2]byte
+	binary.BigEndian.PutUint16(serLen[:], uint16(len(serPub)))
+	ser = append(ser, serLen[:]...)
+	ser = append(ser, serPub...)
+
+	return ser, nil
+}
+
+// unmarshalCredential decodes serialized bytes and returns a credential, if possible.
+func unmarshalCredential(ser []byte) (*Credential, error) {
+	if len(ser) < 8 {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	validTime := time.Duration(binary.BigEndian.Uint32(ser)) * time.Second
+	pubAlgo := SignatureScheme(binary.BigEndian.Uint16(ser[4:6]))
+	pubLen := binary.BigEndian.Uint16(ser[6:8])
+
+	pubKey, err := x509.ParsePKIXPublicKey(ser[8:])
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ser[8:]) != int(pubLen) {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	return &Credential{validTime, pubAlgo, pubKey}, nil
+}
+
+// getCredentialLen returns the number of bytes comprising the serialized
+// credential struct inside the Delegated Credential.
+func getCredentialLen(ser []byte) (int, error) {
+	if len(ser) < 8 {
+		return 0, errors.New("tls: delegated credential is not valid")
+	}
+
+	// The validity time.
+	ser = ser[4:]
+
+	// The expCertVerfAlgo.
+	ser = ser[2:]
+
+	// The lenght of the Public Key.
+	pubLen := int(binary.BigEndian.Uint16(ser))
+	ser = ser[2:]
+
+	if len(ser) < pubLen {
+		return 0, errors.New("tls: delegated credential is not valid")
```

Comment:
Perhaps explain why it is invalid in the error message

Reply:
Mmm.. I didn't wanted to leak that information, and just always state that is invalid.

Reply:
Why would that be leaking information, these errors don't end up on the wire right? This will just be an internal error alert that's communicated outside.